zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+//! Functions for wallet support of ephemeral transparent addresses.
+use std::cmp::max;
+use std::collections::HashMap;
+use std::ops::RangeInclusive;
+
+use rusqlite::{named_params, OptionalExtension};
+
+use zcash_client_backend::{data_api::Account, wallet::TransparentAddressMetadata};
+use zcash_keys::{
+    encoding::{encode_transparent_address_p, AddressCodec},
+    keys::AddressGenerationError,
+};
+use zcash_primitives::{
+    legacy::{
+        keys::{EphemeralIvk, NonHardenedChildIndex, TransparentKeyScope},
+        TransparentAddress,
+    },
+    transaction::TxId,
+};
+use zcash_protocol::consensus;
+
+use crate::{error::SqliteClientError, wallet::get_account, AccountId, SqlTransaction, WalletDb};
+
+/// The number of ephemeral addresses that can be safely reserved without observing any
+/// of them to be mined. This is the same as the gap limit in Bitcoin.
+pub(crate) const GAP_LIMIT: i32 = 20;
+
+// The custom scope used for derivation of ephemeral addresses.
+//
+// This must match the constant used in
+// `zcash_primitives::legacy::keys::AccountPubKey::derive_ephemeral_ivk`.
+//
+// TODO: consider moving this to `zcash_primitives::legacy::keys`, or else
+// provide a way to derive `ivk`s for custom scopes in general there, so that
+// the constant isn't duplicated.
+const EPHEMERAL_SCOPE: TransparentKeyScope = match TransparentKeyScope::custom(2) {
```

Comment:
I would add a const fn `TransparentKeyScope::custom_unchecked` that asserts on invalid values and use that here.

Reply:
Also this should probably be in `zcash_client_backend` as a public constant, rather than buried as a private value in the guts of `zcash_client_sqlite`.

Reply:
That's what the TODO is for.